xapian-applications/omega/handler_libarchive.cc

```diff
@@ -56,68 +268,31 @@ extract(const string& filename,
 	    return false;
 	}
 
-	size_t total;
-	ssize_t size;
-	string s, s_content, s_styles;
-	struct archive_entry* entry;
-
-	// extracting data from content.xml, styles.xml and meta.xml
-	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
-	    string pathname = archive_entry_pathname(entry);
-	    if (pathname == "content.xml") {
-		total = archive_entry_size(entry);
-		s_content.resize(total);
-		size = archive_read_data(archive_obj, &s_content[0], total);
-
-		if (size <= 0) {
-		    error = "Libarchive was not able to extract data from "
-			    "content.xml";
-		    return false;
-		}
-
-		s_content.resize(size);
-	    } else if (pathname == "styles.xml") {
-		total = archive_entry_size(entry);
-		s_styles.resize(total);
-		size = archive_read_data(archive_obj, &s_styles[0], total);
-
-		if (size <= 0) {
-		    error = "Libarchive was not able to extract data from "
-			    "styles.xml";
-		    return false;
-		}
-
-		s_styles.resize(size);
-	    } else if (pathname == "meta.xml") {
-		total = archive_entry_size(entry);
-		string s_meta;
-		s_meta.resize(total);
-		size = archive_read_data(archive_obj, &s_meta[0], total);
-
-		if (size > 0) {
-		    // indexing file even if this fails
-		    s_meta.resize(size);
-		    MetaXmlParser metaxmlparser;
-		    metaxmlparser.parse(s_meta);
-		    title = metaxmlparser.title;
-		    keywords = metaxmlparser.keywords;
-		    author = metaxmlparser.author;
-		}
-	    }
+	bool succeed;
+	if (startswith(mimetype, "application/vnd.sun.xml.") ||
+	    startswith(mimetype, "application/vnd.oasis.opendocument.")) {
+	    succeed = extract_odf(archive_obj, dump, title, keywords, author,
+				  error);
+	    if (!succeed)
+		return false;
```

Comment:
If you're going to put the returned value in a variable, better to declare the variable when you can assign to it - as this currently is the scope of the variable is greater than needed.

But there isn't really much point putting the result in a variable here - you can just test the result directly - e.g. for the above:

```
if (!extract_odf(...)) {
    return false;
}
```

A boolean return for "success" is pretty common and reads nicely inline - "if we didn't extract the odf then return false" - but if the returned thing is more obscure a variable can provide a way to document what it is.  Perhaps `extract_foo()` returns the number of errors encountered, in which case:

```
int error_count = extract_foo(...);
if (error_count > 0)
    return false;
```

is perhaps more helpful than:

```
if (extract_foo(...) > 0)
    return false;
```

And probably clearer than adding a comment such as:

```
// Check if extract_foo() encountered any errors.
if (extract_foo(...) > 0)
    return false;
```

Reply:
Thank you for pointing this out. I have made the necessary changes.